fix(patch): do not leak zones when wrapping requestAnimationFrame

justindujardin opened the pull request:
# Problem
requestAnimationFrame is often called from within the callback function that is bound to its zone.  This results in every subsequent call to the same function creating and binding to a new zone.

# Proposed Solution
Maintain a simple reference count of function signatures given to requestAnimationFrame, and when the reference count is greater than 0 bind to to the parent zone rather than creating a new one.

Feedback welcome.

Fixes: #143, https://github.com/angular/angular/issues/2912

vicb commented:
@justindujardin Thanks for the PR, I'll try to take a look but I have limited availability until the start mid-August (may be someone else can take a look before).

Please remove the commit that updates the dist folder (when only update generated files when we release a new version). Thanks.

justindujardin commented:
@vicb Sure, snipped that commit out.

The problem was originally noted using angular2 dev preview.  I'll produce a custom build of angular2 that uses my fork of zone.js, but this issue will pretty severely degrade the performance of any app that uses a rAF time loop. If you have guidance on who I could reach out to about this, I'm happy to do so.   

For now, my issue is resolved by a custom angular2 build.

jomeier commented:
@vicb @justindujardin Whats the status of this one?

btford commented:
I'm going to look at this with @caitp next week. Looks like this is about ready to go!

justindujardin commented:
@btford That's great.  Let me know if anything needs to change or is unclear.

btford commented:
@vicb can you take a look at this please?

vicb commented:
@justindujardin could you please squash the commits to ease the review ? (The commit label should rather be feat). I'll review by next Tuesday (not sure to have time tonight). Thanks.

justindujardin commented:
@vicb no problem.

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,58 @@ function patchSetClearFunction(obj, fnNames) {
+ *  4) If the reference count is <= 0, clear the cached function signature and dequeue
+ *     the current task.
+ */
+function patchRepeatingCallbackFunction(obj, fnNames) {
```
`patchRequestAnimationFrame` ? _(Not sure)_

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,58 @@ function patchSetClearFunction(obj, fnNames) {
+    if (delegate) {
+      var ids = {};
+      global.zone[name] = function (fn) {
+        if(fn){
```
minor: `if (fn) {` - white spaces

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,58 @@ function patchSetClearFunction(obj, fnNames) {
+      global.zone[name] = function (fn) {
+        if(fn){
+          // Reference count functions by their toString() value.
+          var fnRef = fn, id = fn.toString();
```
consistency:
```
var fnRef = fn;
var id = fn.toString();
```

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,58 @@ function patchSetClearFunction(obj, fnNames) {
+            ids[id]--;
+            if (ids[id] <= 0) {
+              delete ids[id];
+              zone.dequeueTask();
```
zone.dequeueTask(fnRef);

justindujardin replied:
I'm still a bit unsure about how the `dequeueTask` method is supposed to work, because it is unimplemented in the base zone implementation.  Thanks for the correction.

vicb commented:
@justindujardin I have added (mostly minor) comments in the PR.

Two questions:
- What about adding a property to the callback instead of maintaining an array of ids ?
- Couldn't we only say `patchPrototype(global, ['requestAnimationFrame', 'mozRequestAnimationFrame', 'webkitRequestAnimationFrame']` - the callback would be executed in the zone when the RAF is called.

Maybe the first thing would be to consider (or reject) the last bullet above and if it can not work then re-work the implementation in this PR according to the other comment.

What do you think (you probably have spend much more time than me on this).

justindujardin commented:
@vicb I've applied most of your suggestions, but I don't see how the use of `patchPrototype` would solve this problem. The issue with rAF is that it's called recursively from within a callback that is bound to its own zone. Using `patchPrototype` creates a new child zone for each invocation, which causes the tests to fail. Maybe I'm missing something in what you have suggested.

Your suggestion to track the reference count on the function itself resolves a concern that @btford had about how I was calculating function signatures using `toString()`. :+1:

vicb commented:
> I don't see how the use of patchPrototype would solve this problem ...

you are absolutely right, that was a bad idea.

I'll add an other round of inline comments

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,49 @@ function patchSetClearFunction(obj, fnNames) {
+    var delegate = obj[name];
+    if (delegate) {
+      global.zone[name] = function (fn) {
+        if (fn) {
```
```js
        if (fn) {
          // vicb: this would be called first, moved on top
          if (typeof fn._zonejs_refs === 'undefined') {
            fn._zonejs_refs = 1;
            fn._zonejs_zone = global.zone;       
          } else {
            fn._zonejs_refs++;
          }

          arguments[0] = function () {
            // vicb: moved up (in case the fn throws)
            if (--fn._zonejs_refs === 0) {
              delete fn._zonejs_refs;
            }

            fn.apply(this, arguments);
          };

          // vicb: by using bindOnce dequeue should not be required any more
          arguments[0] = fn._zonejs_zone.bindOnce(arguments[0]);
        }
```

Notes:
- changed the property name to make collision less likely,
- moved up some code so that it match execution order,
- added some comments to explain the changes (should be removed).

Feel free to use it or now

justindujardin replied:
I adopted your changes where possible, but a few subtleties are lost in translation:
 - The decrement cannot happen before the original callback function is executed.  If it does the count will never be > 0 and any rAF calls inside it will not bind to the parent.  
 - In order to ensure the reference is decremented I would need to wrap the `fn.apply` in a try/catch block.  Are there concerns wrapping each callback?

**important**
I also failed to consider the asynchronous recursive function call cases, i.e. what if you call setTimeout -> requestAnimationFrame from within a rAF callback.  The setTimeout call will maintain the zone, because it is patched, and the new call to rAF will have no reference count, and will incorrectly bind a new zone. 

I constructed a simple test case and it demonstrates that the zones grow ad-infinitum in when used as described. 

I have no time to work on this issue now, but I will take a look at it after work is done for the day.  Here's the spec change if you are interested in looking into it.

```javascript
  it('should bind to parent zone when called recursively in an async fashion', function (done) {
    testZone.run(function () {
      var frames = 0;

      function frameCallback() {
        expect(zone).toBeDirectChildOf(testZone);
        if (frames++ > 15) {
          return done();
        }
        setTimeout(function(){
          rAF(frameCallback);
        },1);
      }

      rAF(frameCallback);
    });
  });

```

vicb commented:
I have been thinking about this a little bit more:

We do fork on bind to be able to track stack traces (going up the zone hierarchy). What this PR does it to prevent forking in order to minimize stack usage.

A simpler way to do this would be not to use `zone.bind()` at all and we could use `zone.run()` directly. That being said, we have to keep in mind that the root zone has a special behavior and we probably want to fork when called from there.

What about something like:

```js
function patchRequestAnimationFrame(obj, fnNames) {
  fnNames.forEach(function (name) {
    var delegate = obj[name];
    if (delegate) {
      global.zone[name] = function (fn) {
        var callZone = global.zone.isRoot() ? global.zone : global.zone.fork();
        if (fn) {
          arguments[0] = function () {
            return callZone.run(fn, arguments);
          };
        }
        return delegate.apply(obj, arguments);
      };

      obj[name] = function () {
        return global.zone[name].apply(this, arguments);
      };
    }
  });
};

```

And then we could something in the doc to describe the special behavior of `requestAnimationFrame()` and how to override it thanks to the `requestAnimationFrame` hook ?

@justindujardin @btford what do you think ?

justindujardin commented:
@vicb I like your solution, it is much more direct.  I've updated the patch function to work like you suggest.  :+1: 

The async rAF loop is still a problem, unfortunately.  If you start a rAF loop and use setTimeout, or some other forking patched function, then call rAF from within that callback, it will cause setTimeout to become the culprit in the ever-growing stack problem that kills performance.  I'm not sure what to do about this case.  

I have a test that exhibits the problem behavior commented out at the bottom of the spec file.  Thoughts?

vicb commented on lib/patch/functions.js:
```diff
@@ -48,6 +48,36 @@ function patchSetClearFunction(obj, fnNames) {
 
+
+/**
+ * Patch requestAnimationFrame
```
not sure if this is very useful, the description alone should be enough

vicb commented on test/patch/requestAnimationFrame.spec.js:
```diff
@@ -3,63 +3,82 @@
+      //      if (frames++ > 15) {
+      //        return done();
+      //      }
+      //      setTimeout(function () {
```
Do you see any reason why someone would want to do this ?

I don't and I would delete this test - we can re-open if there is a real use case later on.

Thoughts ?

justindujardin replied:
Sadly, I can easily see someone doing this.  Imagine you want to use requestAnimationFrame but only periodically, so you call setTimeout in your handler and then request the next frame when the handler is triggered.  Sure this would be misguided, but hopefully zone.js wouldn't fail to work seamlessly because you've constructed some questionable code.

I resolved the case by using a hybrid of the two approaches that you have suggested.   I store a reference to single forked zone per callback function.  This satisfies all the tests.

If this isn't satisfactory I can revisit it after work.

vicb replied:
> Sadly, I can easily see someone doing this. Imagine you want to use requestAnimationFrame but only periodically, so you call setTimeout in your handler and then request the next frame when the handler is triggered. Sure this would be misguided, but hopefully zone.js wouldn't fail to work seamlessly because you've constructed some questionable code.

Unless you have a concrete use case in mind, I would revert to the simpler solution. IMO when rAF gets used, you are interested in getting called for each single frame.

justindujardin replied:
Hmm, I don't see the original solution being much simpler, and it definitely fails in this case.

The reason I'm concerned with this issue is that Zone.js is wrapping all code executed in an angular2 app, and given the popularity of angular, I would assume you'll get every kind of bone-headed app setup that is conceivable.  

It's not my project though, so I'll revert this back to what it was originally and remove the test this evening.

vicb replied:
I can also craft a test that fails with your solution when raf is called from 2 different zones. 

So let's come with something that works for most common use cases and adapt later if needed.

vicb commented:
Looks like you have reverted the last changes ???

I was ok with them. The only little missing thing was an update to the README about `zone.rAF, zone.wkRAF, ...` hook, quickly describing the behavior and hook. Something by `By default requestAnimationFrame() does not fork the zone but you can override the behacior using this hook`

Thank you very much for your work on this !

justindujardin commented:
@vicb I'm not sure I understood fully the comments you wanted in the README, please confirm them.  I've reverted to the simpler implementation and removed the async test.

The Travis build seems to be kind of fragile.  It fails intermittently on my commits, with timeouts in unrelated functions on varying operating systems.  I wonder if some of the tests run long enough that slight variations in timings cause them to go over the timeout value?  Have you see this behavior before?

vicb commented:
Great I will merge later today. 

We have seen some timing related Travis failures in the past. It most probably has nothing to do with the changes here. However can open a ticket and link the falling Travis builds,  that would help. 


On August 27, 2015 6:55:52 AM PDT, Justin DuJardin <notifications@github.com> wrote:
>@vicb I'm not sure I understood fully the comments you wanted in the
>README, please confirm them.  I've reverted to the simpler
>implementation and removed the async test.
>
>The Travis build seems to be kind of fragile.  It fails intermittently
>on my commits, with timeouts in unrelated functions on varying
>operating systems.  I wonder if some of the tests run long enough that
>slight variations in timings cause them to go over the timeout value? 
>Have you see this behavior before?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/angular/zone.js/pull/147#issuecomment-135441030

vicb commented:
landed as 3d6dc0850b20fd86eef5879f8283de5ee668a668

That will be very useful, thanks @justindujardin for all your follow up

justindujardin commented:
Thanks for your review, @vicb.  I opened another issue with about the timing sensitive tests.
